### ISSUE TITLE
Support for referencing other files via STEP anchors and references

### ISSUE THREAD

**Issue by Moult** [IfcOpenShell/IfcOpenShell#668]
The STEP ISO 10303 standard allows you to reference other IFC files, and even create a compressed collection of IFC files, see: http://www.steptools.com/stds/step/IS_final_p21e3.html#clause-A-4

In fact, they even give examples of how to do it! It's awesome! I can't believe nobody is doing this! Just search that page for ".ifc" for more cool examples!

Can you please implement support for IfcOpenShell to write an `ANCHOR;` and `REFERENCE;` section, and therefore reference URIs and URI fragments? I can help build in support in Blender as an authoring tool.

**Comment by Moult** [comment]
Here is a small demonstration of an IFC file referencing another one (and providing useful anchors to be referenced by another IFC file):

**project.ifc**

```
ISO-10303-21;
HEADER;
FILE_DESCRIPTION(('ViewDefinition [CoordinationView]'),'2;1');
FILE_NAME('$filename','2019-05-12T12:33:45',('$owner','$email'),('$company'),'IfcOpenShell','IfcOpenShell','');
FILE_SCHEMA(('IFC4'));
ENDSEC;
ANCHOR;
<origin> = #9;
<project> = #30;
ENDSEC;
DATA;
#1=IFCPERSON($,$,'$owner',$,$,$,$,$);
#2=IFCORGANIZATION($,'$company',$,$,$);
#3=IFCPERSONANDORGANIZATION(#1,#2,$);
#4=IFCAPPLICATION(#2,'0.0.1','Blender BIM IFC','118df2cf_ed21_438e_a41');
#5=IFCOWNERHISTORY(#3,#4,$,.ADDED.,$,#3,#4,1557664425);
#6=IFCDIRECTION((1.,0.,0.));
#7=IFCDIRECTION((0.,0.,1.));
#8=IFCCARTESIANPOINT((0.,0.,0.));
#9=IFCAXIS2PLACEMENT3D(#8,#7,#6);
#10=IFCDIRECTION((0.,1.,0.));
#12=IFCDIMENSIONALEXPONENTS(0,0,0,0,0,0,0);
#13=IFCSIUNIT(*,.LENGTHUNIT.,$,.METRE.);
#14=IFCSIUNIT(*,.AREAUNIT.,$,.SQUARE_METRE.);
#15=IFCSIUNIT(*,.VOLUMEUNIT.,$,.CUBIC_METRE.);
#16=IFCSIUNIT(*,.PLANEANGLEUNIT.,$,.RADIAN.);
#17=IFCMEASUREWITHUNIT(IFCPLANEANGLEMEASURE(0.017453292519943295),#16);
#18=IFCCONVERSIONBASEDUNIT(#12,.PLANEANGLEUNIT.,'DEGREE',#17);
#19=IFCUNITASSIGNMENT((#13,#14,#15,#18));
#28=IFCGEOMETRICREPRESENTATIONCONTEXT($,'Model',3,1.E-05,#9,$);
#29=IFCGEOMETRICREPRESENTATIONSUBCONTEXT('Body','Model',*,*,*,*,#28,$,.MODEL_VIEW.,$);
#30=IFCPROJECT('0ovDN8c9v9xABetj5G4Toz',$,'My project',$,$,$,$,(#28),#19);
ENDSEC;
END-ISO-10303-21;
```

and now **building.ifc**

```
ISO-10303-21;
HEADER;
FILE_DESCRIPTION(('ViewDefinition [CoordinationView]'),'2;1');
FILE_NAME('$filename','2019-05-12T12:33:45',('$owner','$email'),('$company'),'IfcOpenShell','IfcOpenShell','');
FILE_SCHEMA(('IFC4'));
ENDSEC;
ANCHOR;
<site> = #4
<building> = #6
<ground_floor> = #8
ENDSEC;
REFERENCE;
#1 = <project.ifc#origin>;
#2 = <project.ifc#project>;
ENDSEC;
DATA;
#3=IFCLOCALPLACEMENT($,#1);
#4=IFCSITE('3K6Po8Fj9CQhOzaTgeZZlU',$,'My site',$,$,#3,$,$,$,$,$,$,$,$);
#5=IFCLOCALPLACEMENT(#3,#1);
#6=IFCBUILDING('0tjzSZIM97xO19qEfAUfEt',$,'My building',$,$,#5,$,$,$,$,$,$);
#7=IFCLOCALPLACEMENT(#5,#1);
#8=IFCBUILDINGSTOREY('0GelCcLR9Cf9zvW$jhWkTS',$,'Ground floor',$,$,#7,$,$,$,$);
#9=IFCRELAGGREGATES('2BzqY$ZaDCtBnXHyoN8zrY',$,$,$,#6,(#8));
#10=IFCRELAGGREGATES('2iMphooOTB2g9POWkmwFtf',$,$,$,#4,(#6));
#11=IFCRELAGGREGATES('2iBQDyzurBMugUedxwQ$FP',$,$,$,#2,(#4));
ENDSEC;
END-ISO-10303-21;
```

... there may also be a root file **ISO-10303.p21** with the following content:

```
ISO-10303-21;
HEADER;
FILE_DESCRIPTION(('ViewDefinition [CoordinationView]'),'2;1');
FILE_NAME('$filename','2019-05-12T12:33:45',('$owner','$email'),('$company'),'IfcOpenShell','IfcOpenShell','');
FILE_SCHEMA(('IFC4'));
ENDSEC;
REFERENCE;
#1 = <project.ifc>;
#2 = <building.ifc>;
ENDSEC;
END-ISO-10303-21;
```

**Comment by aothms** [comment]
Yes, this really is quite awesome, but this comes from ISO 10303-21:2016, at IFC we're a couple versions back. I think we're on the 1994 version, two versions back, but not entirely sure. But yes without the cool stuff.

I think you are aware modularization and federation are really high up the agenda within buildingSMART. And I also think that this 3rd edition provides interesting things to make this happen. Within the IFC community though I think it's seen as too little too late and UML/JSON/OWL are seen as more viable modern alternatives. 

If you haven't done so already share this with the crowd at the buildingSMART forums. It will definitely paint them a picture of what's possible with the modern editions of EXPRESS.

**Comment by Moult** [comment]
Yeah - I figured that they were using an outdated STEP version. That said, perhaps we can implement it anyway as an experimental feature, as it makes sense to use the current STEP version, and other developers too desperately want this feature?

**Comment by Moult** [comment]
Also I can only find references to the STEP spec ISO 10303-21 with no year specified:

https://www.iso.org/standard/70303.html
https://standards.buildingsmart.org/IFC/DEV/IFC4_2/FINAL/HTML/schema/chapter-2.htm

**Comment by Moult** [comment]
Oh, and given that ISO 10303-21 actually _uses_ IFC as an example of how to reference suggests that they support it with IFC...

**Comment by Moult** [comment]
The [bS format page](https://technical.buildingsmart.org/standards/ifc/ifc-formats/) links to [ISO 10303-21:2016](https://www.iso.org/standard/63141.html), which has these features:

```
ISO 10303-21:2016 adds anchor, reference and signature sections to support external references, support for compressed exchange structures in an archive, digital signatures and UTF-8 character encoding.
```

I have updated my code sample above to fix the `HEADER` section. It previously had an implementation level set to `2;1`, whereas this refers to the 1994 ISO standard. The appropriate level to use in `4;2`, as [described here](http://www.steptools.com/stds/step/IS_final_p21e3.html#clause-8-2-2).

**Comment by Moult** [comment]
Here is the equivalent combined IFC file in `2;1` 1994 implementation level, for testing purposes. The output of the following code snippet when parsed should be identical to the 2 (or 3, if you distribute as a compressed bundle) files above:

```
ISO-10303-21;
HEADER;
FILE_DESCRIPTION(('ViewDefinition [CoordinationView]'),'2;1');
FILE_NAME('$filename','2019-05-12T12:33:45',('$owner','$email'),('$company'),'IfcOpenShell','IfcOpenShell','');
FILE_SCHEMA(('IFC4'));
ENDSEC;
DATA;
#1=IFCPERSON($,$,'$owner',$,$,$,$,$);
#2=IFCORGANIZATION($,'$company',$,$,$);
#3=IFCPERSONANDORGANIZATION(#1,#2,$);
#4=IFCAPPLICATION(#2,'0.0.1','Blender BIM IFC','118df2cf_ed21_438e_a41');
#5=IFCOWNERHISTORY(#3,#4,$,.ADDED.,$,#3,#4,1557664425);
#6=IFCDIRECTION((1.,0.,0.));
#7=IFCDIRECTION((0.,0.,1.));
#8=IFCCARTESIANPOINT((0.,0.,0.));
#9=IFCAXIS2PLACEMENT3D(#8,#7,#6);
#10=IFCDIRECTION((0.,1.,0.));
#12=IFCDIMENSIONALEXPONENTS(0,0,0,0,0,0,0);
#13=IFCSIUNIT(*,.LENGTHUNIT.,$,.METRE.);
#14=IFCSIUNIT(*,.AREAUNIT.,$,.SQUARE_METRE.);
#15=IFCSIUNIT(*,.VOLUMEUNIT.,$,.CUBIC_METRE.);
#16=IFCSIUNIT(*,.PLANEANGLEUNIT.,$,.RADIAN.);
#17=IFCMEASUREWITHUNIT(IFCPLANEANGLEMEASURE(0.017453292519943295),#16);
#18=IFCCONVERSIONBASEDUNIT(#12,.PLANEANGLEUNIT.,'DEGREE',#17);
#19=IFCUNITASSIGNMENT((#13,#14,#15,#18));
#28=IFCGEOMETRICREPRESENTATIONCONTEXT($,'Model',3,1.E-05,#9,$);
#29=IFCGEOMETRICREPRESENTATIONSUBCONTEXT('Body','Model',*,*,*,*,#28,$,.MODEL_VIEW.,$);
#30=IFCPROJECT('0ovDN8c9v9xABetj5G4Toz',$,'My project',$,$,$,$,(#28),#19);
#31=IFCLOCALPLACEMENT($,#9);
#32=IFCSITE('3K6Po8Fj9CQhOzaTgeZZlU',$,'My site',$,$,#31,$,$,$,$,$,$,$,$);
#33=IFCLOCALPLACEMENT(#31,#9);
#34=IFCBUILDING('0tjzSZIM97xO19qEfAUfEt',$,'My building',$,$,#33,$,$,$,$,$,$);
#35=IFCLOCALPLACEMENT(#33,#9);
#36=IFCBUILDINGSTOREY('0GelCcLR9Cf9zvW$jhWkTS',$,'Ground floor',$,$,#35,$,$,$,$);
#37=IFCRELAGGREGATES('2BzqY$ZaDCtBnXHyoN8zrY',$,$,$,#34,(#36));
#38=IFCRELAGGREGATES('2iMphooOTB2g9POWkmwFtf',$,$,$,#32,(#34));
#39=IFCRELAGGREGATES('2iBQDyzurBMugUedxwQ$FP',$,$,$,#30,(#32));
ENDSEC;
END-ISO-10303-21;
```

**Comment by Moult** [comment]
Sorry @aothms - this is a bit over my head to implement. Would you have time to take a stab at it?

**Comment by aothms** [comment]
well, I can try to make some time, but the hardest thing for me would be trying to figure out an API to use it. Can you sketch your intended usage of this?

**Comment by Moult** [comment]
Thanks, @aothms - here's an example repo I can imagine:

```
ISO-10303.p21 # root file as per spec
coordinates.ifc
project.ifc
building-foo.ifc
building-bar.ifc
building-foo/architectural/level-0.ifc
building-foo/architectural/level-1.ifc
building-foo/mechanical/level-0.ifc
building-foo/leasing-spaces/level-0.ifc
building-foo/energy-zones/level-0.ifc
building-foo/energy-zones/level-0.gbxml
building-foo/facility-management/warranties.ifc
building-foo/facility-management/product-certificate.ifc
building-foo/facility-management/documents/product.pdf # e.g. for RelAssociatesDocument
building-foo/doors/D01.ifc
building-foo/doors/D02.ifc
building-foo/windows/W01.ifc
building-foo/walls/BLK01.ifc
building-foo/environmental-analysis/concrete-psets.ifc
building-foo/external-materials/concrete.mat # e.g. for ExternallyDefinedSurfaceStyle
building-foo/external-materials/sky/sky.rad # e.g. for ExternallyDefinedSurfaceStyle
```

To read one, here is an initial idea

```
cd my-project-dir/
f = ifcopenshell.open('ISO-10303.p21') # As this opens the root file, all references are recursively iterated through and parsed.
f.by_type('IfcBuilding') # returns the IFCBUILDING defined in building-foo.ifc and building-bar.ifc
f = ifcopenshell.open('building-foo/architectural/level-0.ifc', lazy_loading = True) # just like image lazy-loading on the web, this only parses the references if you ask to resolve an anchor later on
f.by_type('IfcBuilding') # returns [], as this is not defined in this file
x = f.by_type('IfcWall')[0].ReferencedInStructures.RelatingStructure # returns IfcBuildingStorey defined in building-foo.ifc, which is lazy-loaded from an explicit request to resolve an anchor
```

To write one, here's a thought:

```
f = ifcopenshell.open('ISO-10303.p21')
f.create_reference('project.ifc')
f.create_reference('building.ifc')
f = ifcopenshell.open('project.ifc')
project = f.create_entity('IfcProject')
f.create_anchor('project', project)
f = ifcopenshell.open('building.ifc')
site = f.create_entity('IfcSite')
project = f.create_reference('project.ifc', 'project')
f.create_entity('IfcRelAggregates', **{'RelatingObject': project, 'RelatedObjects': [site]})
# So that this does not fail, the data type of f.create_reference needs to be a "mock" type. It would basically mock any IFC element, so there is no strict type checking.
```

I hope it makes sense.

**Comment by aothms** [comment]
> # So that this does not fail, the data type of f.create_reference needs to be a "mock" type. It would basically mock any IFC element, so there is no strict type checking.

`IfcBaseEntity` exists in C++ that is the base class of all entity instances. It would pretty automatically enable this functionality, I think, just need to create a subclass then for ReferencedEntity or something.

> project = f.create_reference('project.ifc', 'project')

Should there be something like an anchor class instead or is strings sufficient?

How about an alternative though. That given two files with instances `F0 { i_0, ...,  i_n }` and `F1 { ... }`. And `F0` references `F1`. And then when one does `F1.create_entity(...., i_n)` it will automatically create anchors and references. Or do you prefer this process more explicit and with more control over anchor names?

**Comment by Moult** [comment]
Awesome, an `IfcBaseEntity` sounds like it might fit the bill!

I think it is necessary to be explicit about anchor names. That way it works just like an "interface" in programming languages. Two parties can agree that their IFC files will provide various anchors with certain agreed upon names, and then people can reference them even before they have been created.

E.g. The client agrees that he will provide the anchor `project`. The architect agrees that he will provide then anchor `building-foo`, `building-foo-ground-floor`, `building-foo-level-1`, etc.

**Comment by aothms** [comment]
> Two parties can agree that their IFC files will provide various anchors with certain agreed upon names

Good point, hadn't thought of that.

**Comment by theoryshaw** [comment]
+1 
Think this would be an exciting feature.

**Comment by Moult** [comment]
By the way, in the latest buildingSMART meeting in Vilnius, it was agreed to move to the latest version of STEP, which would support referencing and anchors. Whoo!

https://github.com/buildingSMART/NextGen-IFC/issues/38
https://github.com/buildingSMART/NextGen-IFC/issues/9

**Comment by JakkoH** [comment]
This is exactly what i'm looking for. Especially when moving to the exploitation phase of construction, it would be wise to know explicitly which 'network' of ifc models create the federated model. Problem is otherwise not to know if the federated model is complete or which models might be impacted by each other. 

I searched a bit, but currently there is no update on this yet?

**Comment by aothms** [comment]
I have been discussing this informally with others in the community and the main issue people see with this approach is that it requires naming anchors specifically. This works well for the case presented here such as building storeys of which there are a well-defined handful. But does not work so well for other federated model use case such as attaching a propertyset to nearly every element in the model. I find it still rather file-centric (you still don't know if the network of files is up to date for example) and don't look forward to the syntax changes. So far I've not found the motivation to implement this.

**Comment by JakkoH** [comment]
Thx for your response Thomas, i see your point. Are there other solutions which ifc provides to relate ifc files to eachother, atleast to kind of know which models are in the federated model? As far as i can come up with otherwise is the IfcDocumentReference. Another option would be using an other system to gather track of the related models, e.g. an ontology describing it, however you would never know it when you're looking at a certain ifc file itself.

**Comment by aothms** [comment]
I think that last option is the most realistic. Like  Information Container
for Data Drop ICDD for example.

Sent from a mobile device. Excuse my brevity. Kind regards, Thomas

Op do 18 feb. 2021 12:03 schreef JakkoH <notifications@github.com>:

**Comment by Moult** [comment]
The main advantage of the STEP anchors/references is that it allows data be referenced from different files, which prevents duplication for both efficiency and to prevent duplication errors. This request is not to do with model federation, which to me is a separate, but much simpler issue (but is solved by coincidence via the `ISO-10303.p21` root file in this proposal). Most federated models do not reference - they duplicate, and are basically nothing more than a list of files, some of which may have coordinate offsets, which could be described as:

```
[
    {"file": "/path/to/a.ifc", "matrix": [...]},
    {"file": "/path/to/b.ifc", "matrix": [...]},
]
```

I'm not sure what other metadata would be useful (e.g. things like author, discipline are already in the IFC).

What did you have in mind @JakkoH ?

**Comment by TZwielehner** [comment]
This kind of mechanism is also useful for 4D and 5D applications (e.g. Bexel). Imagine I get a model, create all tasks etc. I simply want to deliver "my part" as new model and edit the architects' etc. model. Also, I don't want to create many "schedule files" but only one. So some kind of namespacing (file_ID.step_ID) for
relationships could come in handy. Referencing anchors to individual objects in a different ifc file might work for storeys etc. but not when all elements need to be addressed.

**Comment by myoualid** [comment]
Interested in helping to implement and test this out. 

When reading the trail of information across issues, the BS forum, and IFC nextGEN repo issues, it seems everything stopped in 2020 . 

I can't seem to get the current state : where are we at ? have we given up on this ?  ping @Moult @aothms

**Comment by Moult** [comment]
My impression is that everything stopped and USD's federates serialisation makes this redundant.

Dion Moult

Sent from [Proton Mail](https://proton.me/mail/home) for Android.

-------- Original Message --------

**Comment by aothms** [comment]
We did discuss this with some of the other implementers in the community and in the end it wasn't really seem as a sufficient way forward towards collaborative or incremental authoring: It requires upfront coordination on what you expose as anchors which was seen as cumbersome. It doesn't really allow for partial data, nor deletes. It doesn't really play well with the relation heavy data model and some of the per-file constraints we have in the schema.

In the end it's still uncertain whether USD-layer-style composition is going to win, and therefore out of academic interest it's still interesting to review what is discussed here, but if I read the community well I don't think this has a high chance of becoming the basis for a more modern IFC.

**Comment by theoryshaw** [comment]
Man, i personally was looking forward to hacking on this to tease it out. 
It would be nice to find a mechanism to remedy file bloat and duplication. 
Would be nice to abandon on empirical evidence of failure, verses theoretical objections to completeness--an angle the OSArch community is uniquely positioned to tease out.

**Comment by myoualid** [comment]
Agreed that this is still worth pursuing. 

Many of us have started using native IFC through Bonsai. With our educational efforts on Bonsai and IfcOpenShell, this snowball is only growing. 

1: Bloat: Split files per spatial hierarchy. Reconstruct files if necessary.
2: Make 4D and 5D actually work. No merging. instead of merging files or going through hoolahoops to get physical quantities from different consultant files.
3. Project coordination: Geo Location + Spatial hierarchy in one IFC as base for all stakeholders, + the rest split as required by whatever BIM document is required.

... I don't want to make this any longer than it needs to be.

**Comment by myoualid** [comment]
I really want to start this ! It's exciting ! 

I also see a world where USD handles scene composition, while IFC4 remains a source for authoring building data: 

- USD federates layers and composes the 3D scenes. It references external IFC data sources (many many step files!). 
- IFC as standalone data source. Native editing of IFC. Generating mesh geometry when required and pulling data on demand.

--> USD scene stays in sync with IFC, through a convention to regenerate geometry after IFC edits; similar to what we do in Bonsai ! 

I am exploring this venue for the web app we're building at aeco dev. Let's see if it fails or succeeds!

**Comment by aothms** [comment]
Likely part of the reason why nobody started this is that it's fairly high complexity in terms of implementation (update the grammar, handle cross-file-boundary references) although there is nothing fundamentally blocking this in ifcopenshell C++ (instance references are resolved lazily so additional file sources could always be loaded) it does have a lot of moving parts. But also high complexity in UX, how to make sure you're creating instances in the correct file, how to package the federated data, etc., but those aspects would also apply to a USD-style federation.

As always, I'm happy to help, but my free time budget is pretty much already spent as I'm currently rewriting most of the C++ code towards a v1.0 data model to finally get the C++ code in a useable state and solve the crashes on referencing deleted instances.

**Comment by aothms** [comment]
> an angle the OSArch community is uniquely positioned to tease out

Yeah... "Every industry gets the serialisation format they deserve".

In a more data-driven and proficient industry people would have already applied all sorts of formats to IFC, such as protobuf, FlatBuffers, Cap’n Proto, Apache Avro, Thrift, ASN.1 (DER, CER, PER, BER), CBOR, MessagePack, BSON, Amazon Ion, Arrow IPC, Parquet.

But in our field we're stuck to the Express version from 2002.

It's a bit tiring when everything needs to come from only a handful of people.

**Comment by theoryshaw** [comment]
I do feel that handful of people is slowly growing however, due in part, as @myoualid mentions, _"our educational efforts on Bonsai and IfcOpenShell"_.

The question: Does this handful grow more from extension, or reinvention? 

If the evolution, and sorted past of html, javascript, sql, csv, email, xml are any indicator, extension seems to make more sense, at times.
